### PR TITLE
Display rarity for each trait in NFTs from OpenSea API #4053

### LINKED ISSUE #4053: Display rarity for each trait in NFTs from OpenSea API
https://github.com/AlphaWallet/alpha-wallet-ios/issues/3730#issuecomment-1062513311

Would you help buzz Tomek if you need design help for it?

Screenshot 2022-03-09 at 3 28 01 PM

### PR DESCRIPTION
**Pull request by oa-s**
Closes #4053

### CODE REVIEW

File: AlphaWallet/Localization/zh-Hans.lproj/Localizable.strings [hunk 1]
```diff
@@ -743,6 +743,7 @@
 "nonfungibles.value.blockchain" = "Blockchain";
 "nonfungibles.value.tokenStandard" = "Token Standard";
 "nonfungibles.value.contract" = "Contract";
+"nonfungibles.value.rariry" = " %@% have this trait";
```
Comment (hboon) [inline AlphaWallet/Localization/zh-Hans.lproj/Localizable.strings:746]: Typo in "rariry"

**Review by hboon — COMMENTED** [review]
Doesn't show up when I look at https://opensea.io/assets/0xff9c1b15b16263c61d017ee9f65c50e4ae0113d7/5913.

`traitsToDisplay` is empty in https://github.com/oa-s/alpha-wallet-ios/blob/ae4e13d8938313b2919bb6889b23fe74addafd58/AlphaWallet/Tokens/Helpers/TokenInstanceViewConfigurationHelper.swift#L144

**Comment by oa-s** [comment]
Rarity might be missed due to request throttling, for some reasons even if we send request with 2 sec delay, it still get throttled by Open Sea. The Opensea API has 1 sec throttled between requests. And even with it, it failures.

**Comment by oa-s** [comment]
[JamesSmartCell](https://github.com/JamesSmartCell) did you have some issues with Opensea request throttling?

**Comment by hboon** [comment]
> Rarity might be missed due to request throttling

@oa-s can you help to add some logging for that with `infoLog("[OpenSea] Rated limited: xxx")`? Maybe check for `429`s? Easier to spot.

**Comment by oa-s** [comment]
> > Rarity might be missed due to request throttling
> 
> @oa-s can you help to add some logging for that with `infoLog("[OpenSea] Rated limited: xxx")`? Maybe check for `429`s? Easier to spot.

done

**Comment by hboon** [comment]
Sorry, I made a typo. Can you help fix it? From:

```
infoLog("[OpenSea] Rated limited: \(url)")
```

to: 

```
infoLog("[OpenSea] Rate limited: \(url)")
```

When not encountering the rate limit, I see:

this|and this
-|-
Screenshot 2022-03-22 at 3 48 30 PM|Screenshot 2022-03-22 at 3 48 32 PM

but still don't see the properties and rarity:

Screenshot 2022-03-22 at 3 48 01 PM

It's this NFT token ID: https://opensea.io/assets/0xff9c1b15b16263c61d017ee9f65c50e4ae0113d7/5913

I see it for this other one though:

Screenshot 2022-03-22 at 3 51 02 PM

---

Just realised we don't have the UI design to display rarity for properties. @colourfreak, can you help to come up with something?

Screenshot 2022-03-22 at 3 51 02 PM

This is how it looks like on OpenSea:

Screenshot 2022-03-22 at 3 48 01 PM

- [ ] Buzz Android when UI design is ready https://github.com/AlphaWallet/alpha-wallet-android/issues/2440

**Comment by hboon** [comment]
> Rarity might be missed due to request throttling, for some reasons even if we send request with 2 sec delay, it still get throttled by Open Sea. The Opensea API has 1 sec throttled between requests. And even with it, it failures.

The OpenSea API key is committed into the repo, so that's the one that is getting rate limited.

I'll send you another one privately.

**Comment by hboon** [comment]
@oa-s this one's ready?

**Comment by hboon** [comment]
@oa-s is this design used? https://github.com/AlphaWallet/alpha-wallet-ios/issues/4136#issuecomment-1078625521

**Comment by oa-s** [comment]
PR implements this layout for traits
https://github.com/AlphaWallet/alpha-wallet-ios/issues/4136#issuecomment-1077373188
do i need to update it to use this one? https://github.com/AlphaWallet/alpha-wallet-ios/issues/4136#issuecomment-1078625521

**Comment by hboon** [comment]
> PR implements this layout for traits
> [#4136 (comment)](https://github.com/AlphaWallet/alpha-wallet-ios/issues/4136#issuecomment-1077373188)
> do i need to update it to use this one? [#4136 (comment)](https://github.com/AlphaWallet/alpha-wallet-ios/issues/4136#issuecomment-1078625521)

Never mind. That's good too. Ready for review and merge?

**Comment by oa-s** [comment]
> Ready for review and merge?

yep

**Review by hboon — APPROVED** [review]
Looks great. Thanks!

**Pull request merged**